src/Storages/MergeTree/ReplicatedMergeTreeMergeStrategyPicker.cpp

```diff
@@ -0,0 +1,151 @@
+    : storage(storage_)
+{}
+
+
+bool ReplicatedMergeTreeMergeStrategyPicker::isMergeFinishedByReplica(const String & replica, const ReplicatedMergeTreeLogEntryData & entry)
+{
+    /// those have only seconds resolution, so recheck period is quite rough
+    auto reference_timestamp = entry.last_postpone_time;
+    if (reference_timestamp == 0)
+        reference_timestamp = entry.create_time;
+
+    /// we don't want to check zookeeper too frequent
+    if (time(nullptr) - reference_timestamp >= RECHECK_MERGE_READYNESS_INTERVAL_SECONDS)
+    {
+        return storage.checkReplicaHavePart(replica, entry.new_part_name);
+    }
+
+    return false;
+}
+
+
+bool ReplicatedMergeTreeMergeStrategyPicker::shouldMergeOnSingleReplica(const ReplicatedMergeTreeLogEntryData & entry) const
+{
+    time_t threshold = execute_merges_on_single_replica_time_threshold;
+    return (
+        threshold > 0       /// feature turned on
+        && entry.type == ReplicatedMergeTreeLogEntry::MERGE_PARTS /// it is a merge log entry
+        && entry.create_time + threshold > time(nullptr)          /// not too much time waited
+    );
+}
+
+
+/// that will return the same replica name for ReplicatedMergeTreeLogEntry on all the replicas (if the replica set is the same).
+/// that way each replica knows who is responsible for doing a certain merge.
+
+/// in some corner cases (added / removed / deactivated replica)
+/// nodes can pick different replicas to execute merge and wait for it (or to execute the same merge together)
+/// but that doesn't have a significant impact (in one case it will wait for the execute_merges_on_single_replica_time_threshold,
+/// in another just 2 replicas will do the merge)
+std::optional<String> ReplicatedMergeTreeMergeStrategyPicker::pickReplicaToExecuteMerge(const ReplicatedMergeTreeLogEntryData & entry)
+{
+    /// last state refresh was too long ago, need to sync up the replicas list
+    if (time(nullptr) - last_refresh_time > REFRESH_STATE_MAXIMUM_INTERVAL_SECONDS)
+        refreshState();
+
+    auto hash = getEntryHash(entry);
+
+    std::lock_guard lock(mutex);
+
+    auto num_replicas = active_replicas.size();
+
+    if (num_replicas == 0)
+        return std::nullopt;
+
+    auto replica_index = static_cast<int>(hash % num_replicas);
+
+    if (replica_index == current_replica_index)
+        return std::nullopt;
+
+    return active_replicas.at(replica_index);
+}
+
+
+void ReplicatedMergeTreeMergeStrategyPicker::refreshState()
+{
+    auto threshold = storage.getSettings()->execute_merges_on_single_replica_time_threshold.totalSeconds();
+
+    auto now = time(nullptr);
+
+    /// the setting is not changed, and last state refresh was done recently
+    if (threshold == execute_merges_on_single_replica_time_threshold
+        && now - last_refresh_time < REFRESH_STATE_MINIMUM_INTERVAL_SECONDS)
+        return;
+
+    if (threshold == 0)
```

Comment:
We can exit earlier if the feature just disabled?

Reply:
:+1: